# Proyecto reconocimiento de estado de ánimo.

Universidad Autónoma de Yucatán  
Facultad de ingeniería  
Aprendizaje profundo para visión por computadora

> **Integrantes:**
* Lugo Canul Sebastian Antonio
* Villanueva Pavia Gustavo Ivan

## Objetivo del proyecto
* Detectar el rostro de una persona y mostrar su estado
* Deberá probar su código en 10 imágenes con estado positivo y 10 con estado negativo
* Probar al menos dos de los modelos de clasificación vistos en esta unidada (SVM, DT, RF)

Las imágenes son tipo JPG de 64 x 64 pixéles

## Instrucciones
1. Este ejercicio puede ser en equipo de hasta 4 integrantes
2. Utiliszar las imágenes de la carpeta **datos/faces** para la obtención de la base de datos
3. Utilizar las notebook **DL_12_Parte1_ClasificadorHaarCascade** y **DL_12_Parte2_DataFrame-Imagenes** como apoyo a la solución del problema
4. Subir el archivo **img_faces.csv** a un repositorio de algún integrante del equipo de **Github**
5. Deberá crear una notebook de Google Colab con el nombre **DL_12_Apellido1_Apellido2_Apellido3_Apellido4.ipynb**.
6. En la notebook deberá incluir una conclusión de los resultados obtenidos y de los retos que implicó este problema.
7. Subir a la plataforma la notebook de Colab y compartirla (enrique.camacho@gmail.com)

## Librerias a usar


In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import cv2 as cv
import os

## Creación de datframes de training y test

Subir ZIP DL_12_EstadosDeAnimo.zip directamente (ya que pesa lo suficiente para que Github no permita subirlo)

In [ ]:
from google.colab import files

In [ ]:
files.upload()

Saving DL_12_EstadosDeAnimo.zip to DL_12_EstadosDeAnimo.zip


In [ ]:
!unzip -q "DL_12_EstadosDeAnimo.zip"

replace DL_12_EstadosDeAnimo/datos/df_faces_test.csv? [y]es, [n]o, [A]ll, [N]one, [r]ename: n
replace DL_12_EstadosDeAnimo/datos/faces/negatives/10.jpg? [y]es, [n]o, [A]ll, [N]one, [r]ename: n
replace DL_12_EstadosDeAnimo/datos/faces/negatives/10000.jpg? [y]es, [n]o, [A]ll, [N]one, [r]ename: A


In [ ]:

!ls


DL_12_EstadosDeAnimo  DL_12_EstadosDeAnimo.zip	sample_data


## Creación del dataframe

In [ ]:
!ls DL_12_EstadosDeAnimo/datos/faces

negatives  positives


In [ ]:


train_positive = os.path.join('./DL_12_EstadosDeAnimo/datos/faces/positives/')

train_negative = os.path.join('./DL_12_EstadosDeAnimo/datos/faces/negatives/')

test_positive = os.path.join('./DL_12_EstadosDeAnimo/datos/faces_test/negatives/')

test_negative = os.path.join('./DL_12_EstadosDeAnimo/datos/faces_test/negatives/')

In [ ]:
print(f'training positivos: {len(os.listdir(train_positive))}')
print(f'training negativos: {len(os.listdir(train_negative))}')
print(f'test positivos: {len(os.listdir(test_positive))}')
print(f'test negativos: {len(os.listdir(test_negative))}')

training positivos: 3695
training negativos: 9476
test positivos: 6
test negativos: 6


Los resultados fueron:
* Train positive: 3695
* Train negativos: 9476
* Test positivos:6
* Test negativos:6

## Cantidad de imágenes por Directorio

In [ ]:
from glob import glob

In [ ]:
df_positive = pd.DataFrame()
for imagen in glob(f'{train_positive}*.jpg'):
    img = cv.imread(imagen)
    img_gray = cv.cvtColor(img, cv.COLOR_BGR2GRAY)
    img_r = cv.resize(img_gray, (40,40), interpolation = cv.INTER_NEAREST)
    fila = np.append(1, img_r.flatten())
    fila = fila.reshape(1, 1601  )
    df_positive = pd.concat([df_positive,  pd.DataFrame(fila)  ])
print('Tamaño de dataframe: ',df_positive.shape)

Tamaño de dataframe:  (3690, 1601)


In [ ]:
df_positive.head()

0     1     2     3     4     5     6     7     8     9     ...  1591  \
0     1   119   118   121   120   125   116   114   128   137  ...    30   
0     1    78    86   101   110   125   136   139   144   145  ...   115   
0     1   155   152   148   152   160   164   167   170   171  ...   136   
0     1    99    98    98   101   121   162   163   168   170  ...   175   
0     1    18    16    16    15    18    40    44    37    41  ...   126   

   1592  1593  1594  1595  1596  1597  1598  1599  1600  
0    19     8     8     6    10    10    12    12    11  
0   116   111   111   106    84    74    68    68    55  
0   143   130    83    16    27    31    59    67    64  
0   176   164   155   145   128   125   125   125   126  
0   124   118   119   123   123   119   122   121   123  

[5 rows x 1601 columns]

In [ ]:
df_negative = pd.DataFrame()
for imagen in glob(f'{train_negative}*.jpg'):
    img = cv.imread(imagen)
    img_gray = cv.cvtColor(img, cv.COLOR_BGR2GRAY)
    img_r = cv.resize(img_gray, (40,40), interpolation = cv.INTER_NEAREST)
    fila = np.append(0, img_r.flatten())
    fila = fila.reshape(1, 1601  )
    df_negative = pd.concat([df_negative,  pd.DataFrame(fila)  ])
print('Tamaño de dataframe: ',df_negative.shape)

Tamaño de dataframe:  (9475, 1601)


In [ ]:
df_negative.head()

0     1     2     3     4     5     6     7     8     9     ...  1591  \
0     0   117   119   121   115   141   163   161   174   174  ...   124   
0     0   129   136   140   140   137   134   138   144   140  ...    88   
0     0   191   190   170   148   128   135   131   133   136  ...   181   
0     0   197   224   232   235   221   195   182   167   165  ...    69   
0     0   126   120   121   123   129   137   140   146   147  ...    99   

   1592  1593  1594  1595  1596  1597  1598  1599  1600  
0   117   100    97    99   105   108   118   118   109  
0    98   107   110   107   111   111   109   112   120  
0   180   171   163   152   137   130   115   110   105  
0    70    72    72    73    73    75    79    84    29  
0    96    89    88    76    70    66    58    56    53  

[5 rows x 1601 columns]

In [ ]:
frames = [df_positive, df_negative]
df_faces = pd.concat(frames)
df_faces.head()

0     1     2     3     4     5     6     7     8     9     ...  1591  \
0     1   119   118   121   120   125   116   114   128   137  ...    30   
0     1    78    86   101   110   125   136   139   144   145  ...   115   
0     1   155   152   148   152   160   164   167   170   171  ...   136   
0     1    99    98    98   101   121   162   163   168   170  ...   175   
0     1    18    16    16    15    18    40    44    37    41  ...   126   

   1592  1593  1594  1595  1596  1597  1598  1599  1600  
0    19     8     8     6    10    10    12    12    11  
0   116   111   111   106    84    74    68    68    55  
0   143   130    83    16    27    31    59    67    64  
0   176   164   155   145   128   125   125   125   126  
0   124   118   119   123   123   119   122   121   123  

[5 rows x 1601 columns]

In [ ]:
print("Tamaño de dataframe ",df_faces.shape)

Tamaño de dataframe  (13165, 1601)


In [ ]:
from sklearn.utils import shuffle
df_faces = shuffle(df_faces)
df_faces.head()

0     1     2     3     4     5     6     7     8     9     ...  1591  \
0     0   143   133   123   125   114   109   108   106   107  ...   144   
0     0   169   159    94    53    34    29    27    23    33  ...    99   
0     0   102   113   117   125   144   151   158   167   170  ...   106   
0     0    44    62   104   111   132   133   121   102   100  ...   124   
0     0   174   191   209   217   226   227   230   227   226  ...   250   

   1592  1593  1594  1595  1596  1597  1598  1599  1600  
0   135   122   121   120   116   114   106    96    83  
0   102   110   111    97    88    81    69    69    66  
0   108   118   120    89    55    62    95    97   115  
0   121   121   116    64    36    34    33    30    31  
0   251   253   249   162   134   132   130   133   132  

[5 rows x 1601 columns]

## Test

In [ ]:
df_test_negative = pd.DataFrame()
for imagen in glob(f'{test_negative}*.jpg'):
    img = cv.imread(imagen)
    img_gray = cv.cvtColor(img, cv.COLOR_BGR2GRAY)
    img_r = cv.resize(img_gray, (40,40), interpolation = cv.INTER_NEAREST)
    fila = np.append(0, img_r.flatten())
    fila = fila.reshape(1, 1601  )
    df_test_negative = pd.concat([df_test_negative,  pd.DataFrame(fila)  ])
print("Tamaño de dataframe ",df_test_negative.shape)
df_test_negative.head()

Tamaño de dataframe  (6, 1601)


0     1     2     3     4     5     6     7     8     9     ...  1591  \
0     0    65    74    88    91    98   104   105   105   102  ...   145   
0     0    69    73    77    78    67    58    58    68    76  ...   102   
0     0   138   136   133   134   140   144   148   156   158  ...   131   
0     0    59    56    43    39    36    36    41    50    55  ...    64   
0     0   164    96   156   177   183   195   205   215   207  ...    34   

   1592  1593  1594  1595  1596  1597  1598  1599  1600  
0   142   138   139   137   120   117   132   116    92  
0    96   100   109   123   119   120   122   109    80  
0   128   122   123   117   109   105    85   103   149  
0    63    68    70    68    70    74    76    79    71  
0    30    42    49    87    82    84   109   112    90  

[5 rows x 1601 columns]

In [ ]:
df_test_positive = pd.DataFrame()
for imagen in glob(f'{test_positive}*.jpg'):
    img = cv.imread(imagen)
    img_gray = cv.cvtColor(img, cv.COLOR_BGR2GRAY)
    img_r = cv.resize(img_gray, (40,40), interpolation = cv.INTER_NEAREST)
    fila = np.append(1, img_r.flatten())
    fila = fila.reshape(1, 1601  )
    df_test_positive = pd.concat([df_test_positive,  pd.DataFrame(fila)  ])
print("Tamaño de dataframe ",df_test_positive.shape)
df_test_positive.head()

Tamaño de dataframe  (6, 1601)


0     1     2     3     4     5     6     7     8     9     ...  1591  \
0     1    65    74    88    91    98   104   105   105   102  ...   145   
0     1    69    73    77    78    67    58    58    68    76  ...   102   
0     1   138   136   133   134   140   144   148   156   158  ...   131   
0     1    59    56    43    39    36    36    41    50    55  ...    64   
0     1   164    96   156   177   183   195   205   215   207  ...    34   

   1592  1593  1594  1595  1596  1597  1598  1599  1600  
0   142   138   139   137   120   117   132   116    92  
0    96   100   109   123   119   120   122   109    80  
0   128   122   123   117   109   105    85   103   149  
0    63    68    70    68    70    74    76    79    71  
0    30    42    49    87    82    84   109   112    90  

[5 rows x 1601 columns]

In [ ]:
frames_test = [df_test_positive, df_test_negative]
df_test_faces = pd.concat(frames_test)
print("Tamaño de dataframe ",df_test_faces.shape)
df_test_faces.head()

Tamaño de dataframe  (12, 1601)


0     1     2     3     4     5     6     7     8     9     ...  1591  \
0     1    65    74    88    91    98   104   105   105   102  ...   145   
0     1    69    73    77    78    67    58    58    68    76  ...   102   
0     1   138   136   133   134   140   144   148   156   158  ...   131   
0     1    59    56    43    39    36    36    41    50    55  ...    64   
0     1   164    96   156   177   183   195   205   215   207  ...    34   

   1592  1593  1594  1595  1596  1597  1598  1599  1600  
0   142   138   139   137   120   117   132   116    92  
0    96   100   109   123   119   120   122   109    80  
0   128   122   123   117   109   105    85   103   149  
0    63    68    70    68    70    74    76    79    71  
0    30    42    49    87    82    84   109   112    90  

[5 rows x 1601 columns]

In [ ]:
from sklearn.utils import shuffle
df_test_faces = shuffle(df_test_faces)
df_test_faces.head()

0     1     2     3     4     5     6     7     8     9     ...  1591  \
0     1    59    56    43    39    36    36    41    50    55  ...    64   
0     0    65    74    88    91    98   104   105   105   102  ...   145   
0     0   128   137   140   145   149   164   175   187   196  ...   141   
0     1   164    96   156   177   183   195   205   215   207  ...    34   
0     1    69    73    77    78    67    58    58    68    76  ...   102   

   1592  1593  1594  1595  1596  1597  1598  1599  1600  
0    63    68    70    68    70    74    76    79    71  
0   142   138   139   137   120   117   132   116    92  
0   141   128   122    96    79    71    70    72    61  
0    30    42    49    87    82    84   109   112    90  
0    96   100   109   123   119   120   122   109    80  

[5 rows x 1601 columns]

In [ ]:
df_faces.to_csv('img_train_faces.csv', index=False)
df_test_faces.to_csv('img_test_faces.csv', index=False)

Los CSV creados son los que se subirán a la plataforma Github

## Creación y entrenamiento de modelo
Declaración de dataframes desde CSV

In [ ]:
!git clone https://github.com/Sebastian-Lugo/DL12

Cloning into 'DL12'...
fatal: could not read Username for 'https://github.com': No such device or address


In [ ]:
!ls

DL_12			  DL_12_UADY	       sample_data
DL_12_EstadosDeAnimo	  img_test_faces.csv
DL_12_EstadosDeAnimo.zip  img_train_faces.csv


In [ ]:
!ls DL_12.2

faces_csv.rar			     modelo_DT_grid2.pkl  modelo_RF_grid2.pkl
haarcascade_frontalface_default.xml  modelo_DT_grid.pkl   modelo_RF_grid.pkl


In [ ]:
!unrar e DL_12_UADY/faces_csv.rar


UNRAR 5.61 beta 1 freeware      Copyright (c) 1993-2018 Alexander Roshal


Extracting from DL_12_UADY/faces_csv.rar


Would you like to replace the existing file img_test_faces.csv
 76219 bytes, modified on 2023-03-25 01:59
with a new one
 76219 bytes, modified on 2023-03-16 20:19

[Y]es, [N]o, [A]ll, n[E]ver, [R]ename, [Q]uit 
User break

User break


In [ ]:
!ls

DL_12_EstadosDeAnimo	  DL_12_UADY	      img_train_faces.csv
DL_12_EstadosDeAnimo.zip  img_test_faces.csv  sample_data


In [ ]:
path = "img_train_faces.csv"
df_train=pd.read_csv(path)
print(path.split('/')[-1], " shape: ",df_train.shape)
df_train.head()

KeyboardInterrupt: ignored

In [ ]:
path = "img_test_faces.csv"
df_test=pd.read_csv(path)
print(path.split('/')[-1], " shape: ",df_test.shape)
df_test.head()

In [ ]:
# Definir los conjuntos de datos de las muestras de entrenamiento y prueba
X_train=df_train.iloc[:,1:]
y_train=df_train.iloc[:,0]

X_test=df_test.iloc[:,1:]
y_test=df_test.iloc[:,0]

## Decision Tree Classifier

In [ ]:
#DecisionTree pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.tree import DecisionTreeClassifier
from sklearn.pipeline import Pipeline

steps=[
    ('scaler',StandardScaler()),
    ('DecisionTreeClassifier',DecisionTreeClassifier())
]
pipeline=Pipeline(steps)

In [ ]:
#Establecer parámetros
parametros={'DecisionTreeClassifier__max_depth': [4,12,8], 
            'DecisionTreeClassifier__min_samples_leaf':[2,6,10],
            'DecisionTreeClassifier__min_samples_split':[2,3,4]}

In [ ]:
#Realizando tunning
from sklearn.model_selection import GridSearchCV
modelo_DT_grid=GridSearchCV(pipeline, param_grid=parametros, cv=5, n_jobs=-1, verbose=5)

In [ ]:
modelo_DT_grid.get_params()

In [ ]:
#Ajuste
modelo_DT_grid.fit(X_train,y_train)

In [ ]:
modelo_DT_grid.best_params_

In [ ]:
# Encontrar el score
modelo_DT_grid.score(X_test,y_test)

In [ ]:
import joblib
joblib.dump(modelo_DT_grid,'modelo_DT_grid.pkl') # Guardo el modelo.

## RandomForest

In [ ]:
from sklearn.preprocessing import StandardScaler
from sklearn.ensemble import RandomForestClassifier
from sklearn.pipeline import Pipeline

In [ ]:
steps=[
    ('scaler',StandardScaler()),
    ('RandomForestClassifier',RandomForestClassifier())
]
pipeline=Pipeline(steps)

In [ ]:
#Parámetros
parametros={'RandomForestClassifier__max_depth': [4,8,10], 
            'RandomForestClassifier__min_samples_leaf':[2,4,8],
            'RandomForestClassifier__min_samples_split':[2,3,4]}

In [ ]:
from sklearn.model_selection import GridSearchCV
modelo_RF_grid=GridSearchCV(pipeline, param_grid=parametros, cv=5, n_jobs=-1, verbose=5)

In [ ]:
#Ajuste
modelo_RF_grid.fit(X_train,y_train)

In [ ]:
modelo_RF_grid.best_params_

In [ ]:
modelo_RF_grid.score(X_test,y_test)

In [ ]:
import joblib
joblib.dump(modelo_RF_grid,'modelo_RF_grid.pkl') # Guardo el modelo.

En principio, ambos modelos arrojan _scores_ bajos. Sin embargo se ha notado desde la creación de los CSV que existe un desbalance entre los datos positivos y negativos, existen cerca de 4 veces más imágenes negativas que positivas. ¿Mejorará el modelo teniendo un balance entre el número de imágenes positivas y negativas?

## Segundo entrenamiento

In [ ]:
df_negative = df_train[df_train.iloc[:,0] == 0]
df_positive = df_train[df_train.iloc[:,0] == 1]

In [ ]:
print("Hay ",df_negative.shape[0], " imágenes de df_negative")
print("Hay ",df_positive.shape[0], " imágenes de df_positive")

Creemos un balance tal que existan la misma cantidad de imágenes negativas y positivas

In [ ]:
df_negative = df_negative.iloc[:3690,:]

In [ ]:
print("Hay ",df_negative.shape[0], " imágenes de df_negative")
print("Hay ",df_positive.shape[0], " imágenes de df_positive")

In [ ]:
frames = [df_positive, df_negative]
df_train = pd.concat(frames)
df_train.head()

In [ ]:
from sklearn.utils import shuffle
df_train = shuffle(df_train)
print(df_train.shape)
df_train.head()

In [ ]:
# Definir los conjuntos de datos de las muestras de entrenamiento y prueba
X_train=df_train.iloc[:,1:]
y_train=df_train.iloc[:,0]

### Decision Tree Classifier

In [ ]:
from sklearn.preprocessing import StandardScaler
from sklearn.tree import DecisionTreeClassifier
from sklearn.pipeline import Pipeline
from sklearn.model_selection import GridSearchCV

steps=[
    ('scaler',StandardScaler()),
    ('DecisionTreeClassifier',DecisionTreeClassifier())
]
pipeline=Pipeline(steps)

parametros={'DecisionTreeClassifier__max_depth': [6,8,10], 
            'DecisionTreeClassifier__min_samples_leaf':[4,6,8],
            'DecisionTreeClassifier__min_samples_split':[3,5]}

modelo_DT_grid=GridSearchCV(pipeline, param_grid=parametros, cv=5, n_jobs=-1, verbose=5)

modelo_DT_grid.fit(X_train,y_train)

In [ ]:
modelo_DT_grid.best_params_

In [ ]:
modelo_DT_grid.score(X_test,y_test)

In [ ]:
import joblib
joblib.dump(modelo_DT_grid,'modelo_DT_grid2.pkl') # Guardo el modelo.

### RandomForest

In [ ]:
from sklearn.preprocessing import StandardScaler
from sklearn.ensemble import RandomForestClassifier
from sklearn.pipeline import Pipeline
from sklearn.model_selection import GridSearchCV

In [ ]:
steps=[
    ('scaler',StandardScaler()),
    ('RandomForestClassifier',RandomForestClassifier())
]
pipeline=Pipeline(steps)

parametros={'RandomForestClassifier__max_depth': [10,12,14], 
            'RandomForestClassifier__min_samples_leaf':[4,6],
            'RandomForestClassifier__min_samples_split':[3,5]}

modelo_RF_grid=GridSearchCV(pipeline, param_grid=parametros, cv=5, n_jobs=-1, verbose=5)

modelo_RF_grid.fit(X_train,y_train)

In [ ]:
modelo_RF_grid.best_params_

In [ ]:
modelo_RF_grid.score(X_test,y_test)

In [ ]:
import joblib
joblib.dump(modelo_RF_grid,'modelo_RF_grid2.pkl') # Guardo el modelo.

## HAAR Cascade

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import cv2 as cv
import joblib
from google.colab import files

In [ ]:
!git clone https://github.com/Tona1410/DL_12_UADY

In [ ]:
!ls DL_12_UADY

In [ ]:
#Función general de plt.imshow
#La imagen está en blanco y negro
def plt_imshow(img, title:str='',show:bool=True, return_img:bool=False):
    img = cv.cvtColor(img, cv.COLOR_BGR2RGB)    
    plt.imshow(img)
    plt.title(title)
    if show: plt.show()
    if return_img: return img

In [ ]:
#img es la imagen a analizar
#model es el modelo que se usará para detectar el estado de ánimo
#face es una tupla que contiene los parámetros para detectar los rostros em una imagen
#imgTitle y yTitle contienen los títulos de la imágenes
def face_recon(img,model,face:tuple=(1.1,15),
               imgTitle:str='',yTitle:int=0.75):
    #img_original = img.copy()
    img_gray = cv.cvtColor(img, cv.COLOR_BGR2GRAY)

    detector_rostros = cv.CascadeClassifier('DL_12_UADY/haarcascade_frontalface_default.xml')
    faces = detector_rostros.detectMultiScale(img, face[0],face[1])

    for (x,y,w,h) in faces:
        cv.rectangle( img , (x,y),(x+w, y+h), (0,255,0),8 )
        img_face = img_gray[x:x+w,y:y+h]
        try:
          img_face=cv.resize(img_face,(40,40),interpolation = cv.INTER_NEAREST)
          state = model.predict(img_face.reshape((1,1600)))[0]
          if state == 0:
            cv.putText(img,text='Negative :(',org=(x,y+h),fontFace=cv.FONT_HERSHEY_SIMPLEX,fontScale=2,color=(255,255,0),thickness=3, lineType=cv.LINE_AA)
          elif state == 1:
            cv.putText(img,text='Positive :)',org=(x,y+h),fontFace=cv.FONT_HERSHEY_SIMPLEX,fontScale=2,color=(0,255,255),thickness=3, lineType=cv.LINE_AA)
        except: pass
    
    plt_imshow(img)

In [ ]:
model = joblib.load('DL_12_UADY/modelo_DT_grid2.pkl')
model2 = joblib.load('DL_12_UADY/modelo_RF_grid.pkl')

In [ ]:
files.upload()

In [ ]:
! unzip Caras.zip

In [ ]:
img= cv.imread('Caras/negative 1.jpg')
face_recon(img=img,model=model,face=(1.007,10))

In [ ]:
img= cv.imread('Caras/negative 10.jpeg')
face_recon(img=img,model=model2,face=(1.01,15))

In [ ]:
img= cv.imread('Caras/negative 2.jpg')
face_recon(img=img,model=model,face=(1.01,15))

In [ ]:
img= cv.imread('Caras/negative 3.jpg')
face_recon(img=img,model=model2,face=(1.1,15))

In [ ]:
img= cv.imread('Caras/negative 4.jpg')
face_recon(img=img,model=model,face=(1.05,15))

In [ ]:
img= cv.imread('Caras/negative 5.jpg')
face_recon(img=img,model=model2,face=(1.01,15))

In [ ]:
img= cv.imread('Caras/negative 6.jpg')
face_recon(img=img,model=model,face=(1.1,15))

In [ ]:
img= cv.imread('Caras/negative 7.png')
face_recon(img=img,model=model2,face=(1.1,15))

In [ ]:
img= cv.imread('Caras/negative 8.jpeg')
face_recon(img=img,model=model,face=(1.05,15))

In [ ]:
img= cv.imread('Caras/negative 9.jpeg')
face_recon(img=img,model=model2,face=(1.01,15))

In [ ]:
img= cv.imread('Caras/positive 1.jpg')
face_recon(img=img,model=model,face=(1.1,15))

In [ ]:
img= cv.imread('Caras/positive 10.jpg')
face_recon(img=img,model=model2,face=(1.1,15))

In [ ]:
img= cv.imread('Caras/positive 2.jpg')
face_recon(img=img,model=model,face=(1.1,15))

In [ ]:
img= cv.imread('Caras/positive 3.jpg')
face_recon(img=img,model=model2,face=(1.01,15))

In [ ]:
img= cv.imread('Caras/positive 4.jpg')
face_recon(img=img,model=model,face=(1.01,15))

In [ ]:
img= cv.imread('Caras/positive 5.jpg')
face_recon(img=img,model=model2,face=(1.1,15))

In [ ]:
img= cv.imread('Caras/positive 6.jpg')
face_recon(img=img,model=model,face=(1.1,15))

In [ ]:
img= cv.imread('Caras/positive 7.jpg')
face_recon(img=img,model=model2,face=(1.1,15))

In [ ]:
img= cv.imread('Caras/positive 8.jpg')
face_recon(img=img,model=model,face=(1.01,15))

In [ ]:
img= cv.imread('Caras/positive 9.jpg')
face_recon(img=img,model=model2,face=(1.1,15))

## Conclusiones y comentarios

A la hora de resolver la tarea se tuvo en cuenta toda la información dada por el profesor. Había alrededor de 9000 puntos de datos en la carpeta negativa y alrededor de 3000 puntos de datos en la carpeta positiva.

En ambos casos, la puntuación fue del *50 %*, por lo que, si bien hubo casos en los que fue correcta, también hubo casos en los que no lo fue. Parte de los retos fue editar el código en Google Colab, porque en esta plataforma los cambios que hacía un colega no eran visibles para los demás, otro reto era el tiempo de espera para la capacitación, pues el obstáculo era el tiempo de espera. 

Después de hacer pruebas preliminares, notamos que el resultado era del 50 %, y cuando hablábamos entre nosotros, nos preguntamos si la proporción de datos negativos a positivos podría afectar el resultado, porque es una cantidad diferente de datos. Para resolver esta pregunta, intente entrenar los modelos con la misma cantidad de datos positivos y negativos. Sin embargo, la puntuación sigue siendo la misma, por lo que la diferencia en la cantidad de datos no afecta al modelo.